ls1intum/Artemis
# `Notifications`: Add new notification, setting, and email for assessed exercise submissions with scheduling

### Checklist
#### General
#### Server
#### Client

### Motivation and Context
Based on a couple of (old) GitHub issues that want new notification types for multiple different roles and use cases I started implementing them step by step.

This PR adds a new notification (+ settings & email) that is created when an exercises submission has been assessed.

Due to the change request the scheduling logic has to be adapted: 
Now only one scheduled process is created for each exercise (no longer for each submission/assessment) -> the previous decision matrix had to be updated
![image](https://user-images.githubusercontent.com/65814168/147860234-e4697445-7fb9-48c0-9136-650b8948b34d.png)

### Steps for Testing
Prerequisites:
- Use TS1 with your TUM account (needed to check email support)
- 1 Instructor/Tutor/Editor
- 1 Student
- 1 Exercise

Participate in the exercise as a student.
Assess the submission as the other user.

Check that the student received a notification & email (you have to enable the checkbox in your settings first).

It is also important to check if the scheduling part works as intended.
The base cases are:
- send notification/email immediately when the assessment due date is undefined or now/past.
- schedule the notification if the due date is set in the future. This can be tested by setting the due date, e.g. to be in 2-3 minutes, creating a new assessment(or re-assessing the exercise) and waiting a bit to see if the notification does arrive or not.

For the more complicated cases, I have created a decision matrix where you can see what should happen e.g. when you update an exercise (its assessment due date).

@krusche asked me to add this testing step as well:
- Make sure that e.g. if you have 5 students in a course/exercise but only 3 did a valid participation/submission (and 3 have been assessed) that truly only those 3 students will receive a notification/email.

https://user-images.githubusercontent.com/65814168/148947851-14b038fd-6b29-4364-8186-95ff74510c56.mp4


### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage
I added new tests for the respective notification-related parts (just like always (e.g. for SingleUserNotificationService/Factory).
No logic was changed, I simply added one more use case.
-> I merged a couple of reworked tests from https://github.com/ls1intum/Artemis/pull/4555 so I could test the pipeline better (e.g. if emails are send correctly)

### Screenshots
![image](https://user-images.githubusercontent.com/65814168/144489222-c9d06835-3526-47da-a60d-49e510285173.png)
![image](https://user-images.githubusercontent.com/65814168/147859180-0c3c1bae-deb4-4bbf-92b0-accfd1e70e48.png)
![image](https://user-images.githubusercontent.com/65814168/144627157-0cc20582-9b8b-4ff8-a66e-3cc810e3e0bc.png)
